net/url/url.h

```diff
@@ -102,43 +106,157 @@ struct URL {
+      fragment = url.substr(pound_sign_index + 1);
+      url = url.substr(0, pound_sign_index);
+    }
+    const size_t question_mark_index = url.find('?');
+    if (question_mark_index != std::string::npos) {
+      parameters_vector = strings::SplitIntoKeyValuePairs(url.substr(question_mark_index + 1), '=', '&');
+      for (auto& it : parameters_vector) {
+        it.second = DecodeURIComponent(it.second);
+      }
+      parameters.insert(parameters_vector.begin(), parameters_vector.end());
+      url = url.substr(0, question_mark_index);
+    }
+    url_without_parameters = url;
+  }
+
+  std::string operator()(const std::string& key, const std::string& default_value) const {
+    const auto cit = parameters.find(key);
+    if (cit != parameters.end()) {
+      return cit->second;
+    } else {
+      return default_value;
+    }
+  }
+
+  std::string operator[](const std::string& key) const { return operator()(key, ""); }
+
+  bool HasParameter(const std::string& key) const { return parameters.find(key) != parameters.end(); }
+
+  static std::string DecodeURIComponent(const std::string& encoded) {
+    std::string decoded;
+    for (size_t i = 0; i < encoded.length(); ++i) {
+      if (i + 3 <= encoded.length() && encoded[i] == '%') {
+        decoded += static_cast<char>(std::stoi(encoded.substr(i + 1, 2).c_str(), nullptr, 16));
+        i += 2;
+      } else {
+        decoded += encoded[i];
+      }
     }
-    os << host;
-    if (port != DefaultPortForProtocol(protocol)) {
-      os << ':' << port;
+    return decoded;
+  }
+
+  static std::string EncodeURIComponent(const std::string& decoded) {
+    std::string encoded;
+    for (const char c : decoded) {
+      if (::isalpha(c) || ::isdigit(c)) {
+        encoded += c;
+      } else {
+        encoded += strings::Printf("%%%02X", static_cast<int>(c));
+      }
     }
-    os << path;
-    return os.str();
+    return encoded;
   }
 
-  static int DefaultPortForProtocol(const std::string& protocol) {
-    // We don't really support HTTPS/SSL or any other protocols yet -- D.K. :-)
-    return protocol == "http" ? 80 : 0;
+  std::string ComposeParameters() const {
+    std::string composed_parameters;
+    if (!parameters_vector.empty()) {
+      for (size_t i = 0; i < parameters_vector.size(); ++i) {
+        composed_parameters += "?&"[i];
+        composed_parameters += EncodeURIComponent(parameters_vector[i].first) + '=' +
+                               EncodeURIComponent(parameters_vector[i].second);
+      }
+    }
+    if (!fragment.empty()) {
+      composed_parameters += "#" + fragment;
+    }
+    return composed_parameters;
   }
 
-  static std::string DefaultProtocolForPort(int port) { return port == 80 ? "http" : ""; }
+  std::vector<std::pair<std::string, std::string>> parameters_vector;
+  std::map<std::string, std::string> parameters;
+  std::string fragment;
+  std::string url_without_parameters;
 };
 
+struct URL : URLParameters, URLWithoutParameters {
```

Comment:
Logically, URL "is-not-a" URLParameters, it "has-a" URLParameters. The field re-use makes sense.